src/mlpack/methods/xgboost/xgb_exact_numeric_split_impl.hpp

```diff
@@ -0,0 +1,92 @@
+/**
+ * @file methods/xgboost/xgb_exact_numeric_split_impl.hpp
+ * @author Rishabh Garg
+ *
+ * Implementation of exact numeric splitter for XGBoost.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_XGBOOST_XGB_EXACT_NUMERIC_SPLIT_IMPL_HPP
+#define MLPACK_METHODS_XGBOOST_XGB_EXACT_NUMERIC_SPLIT_IMPL_HPP
+
+namespace mlpack {
+namespace ensemble {
+
+template<typename LossFunction>
+template<bool UseWeights, typename VecType, typename MatType,
+          typename WeightVecType>
+double XGBExactNumericSplit<LossFunction>::SplitIfBetter(
+    const double bestGain,
+    const VecType& data,
+    const MatType& input,
+    const WeightVecType& weights,
+    const size_t /* minimumLeafSize */,
+    const double minimumGainSplit,
+    double& splitInfo,
+    AuxiliarySplitInfo& /* aux */,
+    LossFunction& lossFunction)
+{
+  // TODO: Filter out and separate the missing data before sorting to decrease
+  // sensitivity towards missing data.
+
+  // Next, sort the data.
+  arma::uvec sortedIndices = arma::sort_index(data);
+  arma::vec sortedGradients(sortedIndices.n_elem);
+  arma::vec sortedHessians(sortedIndices.n_elem);
+  lossFunction.sortGradAndHess(sortedIndices, sortedGradients,
+      sortedHessians);
+
+  // Sanity check: if the first element is the same as the last, we can't split
+  // in this dimension.
+  if (data[sortedIndices[0]] == data[sortedIndices[sortedIndices.n_elem - 1]])
+    return DBL_MAX;
+
+  // Initialize the binary scan. It initialises the index such that the weight
+  // of the left child is higher than the minChildWeight.
+  size_t index = lossFunction.BinaryScanInitialize(sortedGradients,
+      sortedHessians);
```

Comment:
Do we want to require that loss functions used with `XGBExactNumericSplit` implement the binary splitting strategy?  That's okay with me, but that is something that we should document.  Alternately, you could add a (much slower) overload like in `BestBinaryNumericSplit`.